src/lib.rs

```diff
@@ -1,90 +1,106 @@
-use numpy::ndarray::{ArrayViewD};
-use numpy::PyArray;
-use numpy::{
-    PyArrayDyn, PyReadonlyArrayDyn,
-};
-use pyo3::{
-    pymodule,
-    types::{PyModule},
-    PyResult, Python,
-};
-use std::{f32::consts::PI};
-use num_traits::{Float};
+use anyhow::Result;
+use numpy::PyArray1;
+use pyo3::prelude::*;
+use pyo3::types::*;
+use std::f32::consts::PI;
 
 const TAU: f32 = 2.0 * PI;
 
-fn scale_fn(q: f32, delta: f32) -> f32 {
-    delta / TAU * (2.0 * q - 1.0).asin()
+fn scale_fn(q: f32, delta: f32) -> Result<f32> {
+    Ok(delta / TAU * (2.0 * q - 1.0).asin())
 }
 
-#[derive(Debug)]
+#[pyclass]
 struct TDigest {
     means: Vec<f32>,
-    weights: Vec<i64>
+    weights: Vec<i64>,
 }
 
-
+#[pymethods]
 impl TDigest {
-    fn from_vec(mut vec: Vec<f32>, delta: f32, sort: bool) -> TDigest
-    {
-        if sort {
-            vec.sort_by(|a, b| a.partial_cmp(b).unwrap());
-        }
+    #[getter]
+    fn total_weight(&self) -> PyResult<i64> {
+        Ok(self.weights.iter().sum())
+    }
 
-        let n = vec.len();
-        let percentile_increment: f32 = 1.0 / n as f32;
-        let mut percentile: f32;
-        let mut k_lower = scale_fn(0.0, delta);
-        let mut k_upper: f32;
-        let mut start: usize = 0;
-        let mut sum: f32 = 0.0;
+    fn __len__(&self) -> PyResult<usize> {
+        Ok(self.means.len())
+    }
 
-        let mut means: Vec<f32> = Vec::new();
-        let mut weights: Vec<i64> = Vec::new();
+    fn merge(&mut self, other: &Self) -> PyResult<()> {
+        Ok(())
+    }
 
-        for (i, val) in vec.iter().enumerate() {
-            percentile = (i as f32 + 1.0) * percentile_increment;
-            k_upper = scale_fn(percentile, delta);
-            sum += val;
+    fn quantile(&self, x: f32) -> PyResult<f32> {
```

Comment:
I haven't checked this produces the right numbers. I just translated the python code